File: java/src/test/java/com/consistenthash/ConformanceTest.java
package com.consistenthash;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONObject;
import org.junit.jupiter.api.Test;

/** The Java port must reproduce the committed golden vectors exactly. */
class ConformanceTest {

    private static JSONObject vectors() throws IOException {
        // Walk up from the working directory to the repo root, then into conformance/.
        Path dir = Paths.get("").toAbsolutePath();
        while (dir != null && !Files.exists(dir.resolve("conformance/vectors.json"))) {
            dir = dir.getParent();
        }
        if (dir == null) {
            throw new IOException("could not locate conformance/vectors.json");
        }
        return new JSONObject(Files.readString(dir.resolve("conformance/vectors.json")));
    }

    private static long hex(String s) {
        return Long.parseUnsignedLong(s.substring(2), 16);
    }

    private static List<String> nodes(JSONObject spec) {
        List<String> out = new ArrayList<>();
        spec.getJSONArray("nodes").forEach(o -> out.add((String) o));
        return out;
    }

    @Test
    void hashMatches() throws IOException {
        JSONObject h = vectors().getJSONObject("hash");
        JSONObject h32 = h.getJSONObject("fnv1a_32");
        for (String probe : h32.keySet()) {
            assertEquals(hex(h32.getString(probe)), HashRing.fnv1a32(probe), "fnv32 " + probe);
        }
        JSONObject h64 = h.getJSONObject("fnv1a_64");
        for (String probe : h64.keySet()) {
            assertEquals(hex(h64.getString(probe)), HashRing.fnv1a64(probe), "fnv64 " + probe);
        }
    }

    @Test
    void ringMatches() throws IOException {
        JSONObject spec = vectors().getJSONObject("ring");
        HashRing ring = new HashRing(nodes(spec), spec.getInt("replicas"));
        JSONObject a = spec.getJSONObject("assignment");
        for (String key : a.keySet()) {
            assertEquals(a.getString(key), ring.get(key), "ring " + key);
        }
    }

    @Test
    void rendezvousMatches() throws IOException {
        JSONObject spec = vectors().getJSONObject("rendezvous");
        Rendezvous hrw = new Rendezvous(nodes(spec));
        JSONObject a = spec.getJSONObject("assignment");
        for (String key : a.keySet()) {
            assertEquals(a.getString(key), hrw.get(key), "hrw " + key);
        }
    }

    @Test
    void jumpMatches() throws IOException {
        JSONObject spec = vectors().getJSONObject("jump");
        Jump jump = new Jump(nodes(spec));
        JSONObject a = spec.getJSONObject("assignment");
        for (String key : a.keySet()) {
            assertEquals(a.getString(key), jump.get(key), "jump " + key);
        }
    }

    @Test
    void maglevMatches() throws IOException {
        JSONObject spec = vectors().getJSONObject("maglev");
        Maglev maglev = new Maglev(nodes(spec), spec.getInt("table_size"));
        JSONObject a = spec.getJSONObject("assignment");
        for (String key : a.keySet()) {
            assertEquals(a.getString(key), maglev.get(key), "maglev " + key);
        }
    }

    @Test
    void boundedLoadMatches() throws IOException {
        JSONObject spec = vectors().getJSONObject("bounded_load");
        double eps = (double) spec.getInt("epsilon_num") / spec.getInt("epsilon_den");
        BoundedLoad bounded = new BoundedLoad(nodes(spec), spec.getInt("replicas"), eps);
        JSONObject a = spec.getJSONObject("assignment");
        spec.getJSONArray("keys_in_order").forEach(o -> {
            String key = (String) o;
            assertEquals(a.getString(key), bounded.get(key), "bounded " + key);
        });
    }
}

File: web/src/core/conformance.test.ts
import { describe, expect, it } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";
import {
  BoundedLoad,
  fnv1a32,
  fnv1a64,
  Jump,
  Maglev,
  Rendezvous,
  Ring,
} from "./index";

// The TypeScript port must reproduce the committed golden vectors exactly.
const here = dirname(fileURLToPath(import.meta.url));
const vectors = JSON.parse(
  readFileSync(resolve(here, "../../../conformance/vectors.json"), "utf-8"),
);

const hex = (s: string) => BigInt(s);

describe("conformance", () => {
  it("fnv hash probes", () => {
    for (const [probe, expected] of Object.entries(vectors.hash.fnv1a_32)) {
      expect(BigInt(fnv1a32(probe))).toBe(hex(expected as string));
    }
    for (const [probe, expected] of Object.entries(vectors.hash.fnv1a_64)) {
      expect(fnv1a64(probe)).toBe(hex(expected as string));
    }
  });

  it("ring", () => {
    const spec = vectors.ring;
    const ring = new Ring(spec.nodes, spec.replicas);
    for (const [key, node] of Object.entries(spec.assignment)) {
      expect(ring.get(key)).toBe(node);
    }
  });

  it("rendezvous", () => {
    const spec = vectors.rendezvous;
    const hrw = new Rendezvous(spec.nodes);
    for (const [key, node] of Object.entries(spec.assignment)) {
      expect(hrw.get(key)).toBe(node);
    }
  });

  it("jump", () => {
    const spec = vectors.jump;
    const jump = new Jump(spec.nodes);
    for (const [key, node] of Object.entries(spec.assignment)) {
      expect(jump.get(key)).toBe(node);
    }
  });

  it("maglev", () => {
    const spec = vectors.maglev;
    const maglev = new Maglev(spec.nodes, spec.table_size);
    for (const [key, node] of Object.entries(spec.assignment)) {
      expect(maglev.get(key)).toBe(node);
    }
  });

  it("bounded-load", () => {
    const spec = vectors.bounded_load;
    const eps = spec.epsilon_num / spec.epsilon_den;
    const bounded = new BoundedLoad(spec.nodes, spec.replicas, eps);
    for (const key of spec.keys_in_order) {
      expect(bounded.get(key)).toBe(spec.assignment[key]);
    }
  });
});
